File: ios/Shared/Sync/SyncCrypto.swift
import CommonCrypto
import Foundation

enum SyncCrypto {
    struct KeyBundle: Codable {
        var encryptionKey: Data
        var hmacKey: Data
    }

    static func syncKeyBundle(fromKB kB: Data) -> KeyBundle {
        let info = Data("identity.mozilla.com/picl/v1/oldsync".utf8)
        let material = FxACrypto.hkdf(secret: kB, info: info, length: 64)
        return KeyBundle(
            encryptionKey: Data(material.prefix(32)),
            hmacKey: Data(material.suffix(32))
        )
    }

    static func decryptBSO(payloadJSON: String, keys: KeyBundle) throws -> Data {
        struct Envelope: Decodable {
            let ciphertext: String
            let IV: String
            let hmac: String
        }
        let env = try JSONDecoder().decode(Envelope.self, from: Data(payloadJSON.utf8))
        let actual = FxACrypto.hmacSHA256(key: keys.hmacKey, data: Data(env.ciphertext.utf8))
        // Case-insensitive hex decode preserved from the string-compare era;
        // an unparseable hmac counts as a mismatch.
        let expected = (try? FxACrypto.unhex(env.hmac.lowercased())) ?? Data()
        guard FxACrypto.constantTimeEquals(actual, expected) else { throw SyncError.crypto("bso hmac mismatch") }
        guard let ciphertext = Data(base64Encoded: env.ciphertext),
              let iv = Data(base64Encoded: env.IV)
        else { throw SyncError.crypto("bso base64") }
        // Envelope geometry (SPEC §4): validated after the HMAC so a forged
        // envelope cannot drive decoder buffer sizing.
        guard iv.count == 16 else { throw SyncError.crypto("bso iv length") }
        guard !ciphertext.isEmpty && ciphertext.count % 16 == 0 else { throw SyncError.crypto("bso ciphertext length") }
        return try aes256CBC(decrypt: ciphertext, key: keys.encryptionKey, iv: iv)
    }

    /// Security-framework RNG. A failed copy is a hard error — an all-zero
    /// IV or key from an ignored status would silently break the contract.
    static func secureRandomBytes(_ count: Int) throws -> Data {
        guard count > 0 else { return Data() }
        var bytes = Data(count: count)
        let status = bytes.withUnsafeMutableBytes { buffer in
            SecRandomCopyBytes(kSecRandomDefault, count, buffer.baseAddress!)
        }
        guard status == errSecSuccess else { throw SyncError.crypto("rng failure") }
        return bytes
    }

    static func encryptBSO(plaintext: Data, keys: KeyBundle) throws -> String {
        let iv = try secureRandomBytes(16)
        let cipher = try aes256CBC(encrypt: plaintext, key: keys.encryptionKey, iv: iv)
        let ciphertextB64 = cipher.base64EncodedString()
        let hmac = FxACrypto.hex(FxACrypto.hmacSHA256(key: keys.hmacKey, data: Data(ciphertextB64.utf8)))
        let obj: [String: String] = [
            "ciphertext": ciphertextB64,
            "IV": iv.base64EncodedString(),
            "hmac": hmac
        ]
        return String(data: try JSONSerialization.data(withJSONObject: obj), encoding: .utf8) ?? "{}"
    }

    private static func aes256CBC(encrypt data: Data, key: Data, iv: Data) throws -> Data {
        try crypt(data, key: key, iv: iv, operation: CCOperation(kCCEncrypt))
    }

    private static func aes256CBC(decrypt data: Data, key: Data, iv: Data) throws -> Data {
        try crypt(data, key: key, iv: iv, operation: CCOperation(kCCDecrypt))
    }

    private static func crypt(_ data: Data, key: Data, iv: Data, operation: CCOperation) throws -> Data {
        var out = Data(count: data.count + kCCBlockSizeAES128)
        var moved = 0
        let outCount = out.count
        let status = out.withUnsafeMutableBytes { outBytes in
            data.withUnsafeBytes { inBytes in
                key.withUnsafeBytes { keyBytes in
                    iv.withUnsafeBytes { ivBytes in
                        CCCrypt(
                            operation,
                            CCAlgorithm(kCCAlgorithmAES),
                            CCOptions(kCCOptionPKCS7Padding),
                            keyBytes.baseAddress, key.count,
                            ivBytes.baseAddress,
                            inBytes.baseAddress, data.count,
                            outBytes.baseAddress, outCount,
                            &moved
                        )
                    }
                }
            }
        }
        guard status == kCCSuccess else { throw SyncError.crypto("aes \(status)") }
        out.removeSubrange(moved..<out.count)
        return out
    }
}

File: android/app/src/main/java/de/kjell/zencompanion/sync/SyncCrypto.kt
package de.kjell.zencompanion.sync

import org.json.JSONObject
import java.security.GeneralSecurityException
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

/** Port of `Shared/SyncCrypto.swift` (AES-256-CBC + PKCS7, HMAC over the base64 string). */
object SyncCrypto {
    class KeyBundle(val encryptionKey: ByteArray, val hmacKey: ByteArray)

    fun syncKeyBundle(kB: ByteArray): KeyBundle {
        val info = "identity.mozilla.com/picl/v1/oldsync".toByteArray(Charsets.UTF_8)
        val material = FxACrypto.hkdf(kB, info, 64)
        return KeyBundle(
            encryptionKey = material.copyOfRange(0, 32),
            hmacKey = material.copyOfRange(32, 64),
        )
    }

    fun decryptBSO(payloadJSON: String, keys: KeyBundle): ByteArray {
        val env = JSONObject(payloadJSON)
        val ciphertextB64 = env.getString("ciphertext")
        val ivB64 = env.getString("IV")
        val expectedHmac = try {
            FxACrypto.unhex(env.getString("hmac"))
        } catch (_: SyncError.Crypto) {
            // A non-hex tag can never match; report it as a plain mismatch.
            throw SyncError.Crypto("bso hmac mismatch")
        }
        val actualHmac = FxACrypto.hmacSHA256(keys.hmacKey, ciphertextB64.toByteArray(Charsets.UTF_8))
        // Constant-time comparison so the tag check cannot leak timing.
        if (!MessageDigest.isEqual(actualHmac, expectedHmac)) throw SyncError.Crypto("bso hmac mismatch")
        val ciphertext = b64decode(ciphertextB64)
            ?: throw SyncError.Crypto("bso base64")
        val iv = b64decode(ivB64)
            ?: throw SyncError.Crypto("bso base64")
        // Envelope geometry is validated before decryption (SPEC §4).
        if (iv.size != 16) throw SyncError.Crypto("bso iv length")
        if (ciphertext.isEmpty() || ciphertext.size % 16 != 0) throw SyncError.Crypto("bso ciphertext length")
        return try {
            aes256CBC(ciphertext, keys.encryptionKey, iv, decrypt = true)
        } catch (_: GeneralSecurityException) {
            throw SyncError.Crypto("bso cipher")
        }
    }

    fun encryptBSO(plaintext: ByteArray, keys: KeyBundle): String {
        val iv = ByteArray(16).also { random.nextBytes(it) }
        val cipher = try {
            aes256CBC(plaintext, keys.encryptionKey, iv, decrypt = false)
        } catch (_: GeneralSecurityException) {
            throw SyncError.Crypto("bso cipher")
        }
        val ciphertextB64 = Base64.getEncoder().encodeToString(cipher)
        val hmac = FxACrypto.hex(
            FxACrypto.hmacSHA256(keys.hmacKey, ciphertextB64.toByteArray(Charsets.UTF_8))
        )
        val obj = JSONObject()
        obj.put("ciphertext", ciphertextB64)
        obj.put("IV", Base64.getEncoder().encodeToString(iv))
        obj.put("hmac", hmac)
        return obj.toString()
    }

    private fun aes256CBC(data: ByteArray, key: ByteArray, iv: ByteArray, decrypt: Boolean): ByteArray {
        val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
        cipher.init(
            if (decrypt) Cipher.DECRYPT_MODE else Cipher.ENCRYPT_MODE,
            SecretKeySpec(key, "AES"),
            IvParameterSpec(iv),
        )
        return cipher.doFinal(data)
    }

    internal fun b64decode(s: String): ByteArray? = try {
        Base64.getDecoder().decode(s)
    } catch (_: IllegalArgumentException) {
        try {
            Base64.getMimeDecoder().decode(s)
        } catch (_: IllegalArgumentException) {
            null
        }
    }

    private val random = SecureRandom()
}
